### DISCUSSION TITLE
Unwrap workflow security

### DISCUSSION THREAD

**Discussion by adamgall** [sarcophagus-org/sarcophagus-v2-specs#11] [Ideas]
The "unwrap" workflow for Sarcophagus involves archaeologists publishing private keys to the blockchain. Traditionally, archaeologists used BIP44 seed phrases for all key generation.

https://bitcoin.stackexchange.com/questions/90627

This SO suggests that leaking specific derived data can be used to back-calculate root keys, which can compromise the entire keychain.

Questions:
* is our method safe?
* What assumptions must we put in place to make our method safe?
* Can we make any changes to the archaeologist key derivation workflows to make things safer?
* Are there other sets of leaked data that can be used to back-calculate the root keys? (for example, 1 priv key and 2 pub keys)

This concept needs to be thoroughly understood as we get into V2 development.

**Comment by adamgall** [comment]
See related discussion https://github.com/sarcophagus-org/sarcophagus-v2/discussions/15

**Comment by adamgall** [comment]
The biggest risk that I can find on the internet manifests itself when the following two things are leaked:

1. an `xpub` key at some path like `m/44'/0'/0'`
2. a child `priv` key at some child path like `m/44'/0'/0'/0`

These two things will allow someone to calculate the `xpriv` key at path `m/44'/0'/0'`, which renders _all_ child `priv` keys at path `x` in `m/44'/0'/0'/x` compromised.

A core design decision in Sarcophagus is that archaeologists publish private keys into the public domain.

This is a huge deal, the implications of which could be devastating if our implementations are lazy or incorrectly thought through.

The main question: can an `xpub` key be back-calculated from any of the other data (from the given path `m/44'/0'/0'`) that an archaeologist may leak to the public? Examples of such data would be: multiple child public keys (at `m/44'/0'/0'/x`), multiple child addresses (derived from public keys), multiple on-chain transactions (which expose public keys).

Couple of options:

1. We only use "hardened addresses" for private key generation: https://docs.keys.casa/wealth-security-protocol/rejected-features/hardened-addresses
2. We make sure that archaeologists use different derivation paths (essentially, always a child node of `0`, but different hardened paths before that): https://bitcoin.stackexchange.com/questions/99542. Is this actually different than the first option? Maybe conceptually I'm saying the same thing here.

**Comment by adamgall** [comment]
Clearer understanding of the risks involved are here: https://twitter.com/BTChip/status/1512680818438725633

**Comment by hoylairen** [comment]
yes
